### ISSUE TITLE
[DSIP] [API] Support transfer project ownership.

### ISSUE THREAD

**Issue by CommandCV** [apache/dolphinscheduler#15925]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### Motivation

I created a [tt] project and a workflow definition using the test user, then I want to delete this user using the admin user, but it shows 'Please transform project ownership [tt]', how do I transform project ownership?
![image](https://github.com/apache/dolphinscheduler/assets/49786114/44b6d1ae-a61b-45fd-86f6-3f3311d2269f)

![image](https://github.com/apache/dolphinscheduler/assets/49786114/5dda4b01-d459-483d-ad10-66d7d5a89fad)

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by ruanwenjun** [comment]
Good catch, right now there is no way to transform the project owner, but I think this is not a bug(the error message is incorrect), it should be a feature.

**Comment by niumy0701** [comment]
Hi，There are two options. Let's discuss them together and see which one is better @ruanwenjun @CommandCV 

Option 1 Modify the creator of the associated table
Advantages: No compatibility with historical data, no need to modify permission logic
Disadvantage: There are many modifications to the association table
1. Front end
Add a "Transfer Project" button for user management on the administrator page
(1) Select the corresponding target user

<img width="2038" height="398" alt="Image" src="https://github.com/user-attachments/assets/766edfb7-d174-4a9d-805c-d96e21ff8463" />

(2) Select the projects that need to be transferred

<img width="2030" height="797" alt="Image" src="https://github.com/user-attachments/assets/0b3588e7-1013-4a8f-9a5a-a0d16a4a4a2f" />

(3) Click OK to call the backend interface to transfer the project
2. Backend:
Interface modification of user related table data to ensure data consistency
t_ds_project 
t_ds_project_parameter
t_ds_project_preference
t_ds_relation_project_user
t_ds_workflow_definition 
t_ds_workflow_definition_log
t_ds_task_definition 
t_ds_task_definition_log
t_ds_schedules
t_ds_datasource 
t_ds_relation_datasource_user
t_ds_k8s_namespace
t_ds_relation_namespace_user

Option 2 Increase owner permissions
Advantages: No modification of table associated data
Disadvantage: Requires compatibility with historical data or permission logic compatible with historical logic
1. Front end:
Add the "Grant Owner Permissions" button for authorization

<img width="1598" height="800" alt="Image" src="https://github.com/user-attachments/assets/45f7abf7-76bf-4e77-8d8c-c6640b38f08e" />

2. Backend:
(1) On the basis of the previous permissions, add the owner permission OWNER_PERMISSION=99 to have all permissions. Replace the project creator and modify the verification logic of the previous project creator
(2) Transfer authorization logic before reusing permissions, transfer project permissions, increase owner permissions for target users, and revoke all permissions for source users; The source user and target user belong to the same tenant

**Comment by ruanwenjun** [comment]
@niumy0701 Can we only add a status in User to mark the user is deleted? It's hard to real delete the user, since we need to change a lot of data, e.g. create_user, update_user.

**Comment by niumy0701** [comment]
> [@niumy0701](https://github.com/niumy0701) Can we only add a status in User to mark the user is deleted? It's hard to real delete the user, since we need to change a lot of data, e.g. create_user, update_user.

@ruanwenjun Equivalent to logically deleting a user, how to transfer the projects associated with the user to others for better handling

**Comment by ruanwenjun** [comment]
> > [@niumy0701](https://github.com/niumy0701) Can we only add a status in User to mark the user is deleted? It's hard to real delete the user, since we need to change a lot of data, e.g. create_user, update_user.
> 
> [@ruanwenjun](https://github.com/ruanwenjun) Equivalent to logically deleting a user, how to transfer the projects associated with the user to others for better handling

If we don't really delete the user in DB, then we only need to consider how to change the important resource owner, we can ignore some instance create_user/owner/executor.

**Comment by CommandCV** [comment]
> > > [@niumy0701](https://github.com/niumy0701) Can we only add a status in User to mark the user is deleted? It's hard to real delete the user, since we need to change a lot of data, e.g. create_user, update_user.
> > 
> > 
> > [@ruanwenjun](https://github.com/ruanwenjun) Equivalent to logically deleting a user, how to transfer the projects associated with the user to others for better handling
> 
> If we don't really delete the user in DB, then we only need to consider how to change the important resource owner, we can ignore some instance create_user/owner/executor.

I think we don't really delete the user in DB is a better way, but we still need to support transfer project ownership.

**Comment by CommandCV** [comment]
> Hi，There are two options. Let's discuss them together and see which one is better [@ruanwenjun](https://github.com/ruanwenjun) [@CommandCV](https://github.com/CommandCV)
> 
> Option 1 Modify the creator of the associated table Advantages: No compatibility with historical data, no need to modify permission logic Disadvantage: There are many modifications to the association table
> 
> 1. Front end
>    Add a "Transfer Project" button for user management on the administrator page
>    (1) Select the corresponding target user
> 
> <img alt="Image" width="2000" height="398" src="https://private-user-images.githubusercontent.com/65696095/510589897-766edfb7-d174-4a9d-805c-d96e21ff8463.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI3NzA0NTksIm5iZiI6MTc2Mjc3MDE1OSwicGF0aCI6Ii82NTY5NjA5NS81MTA1ODk4OTctNzY2ZWRmYjctZDE3NC00YTlkLTgwNWMtZDk2ZTIxZmY4NDYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEwVDEwMjIzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4MDVhYzk0MzM2YTY4M2Y4ZTlhMjZjYTllNGVlMzZmYzkzNGE1NGRhY2M4NDVlYjg2YWM1YzUyODI1NmU3ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.q9g61BGDfoU-kuFtE4pYDgM4tqdrzhFSjqnt9KIUSLo">
> (2) Select the projects that need to be transferred
> 
> <img alt="Image" width="2000" height="797" src="https://private-user-images.githubusercontent.com/65696095/510589963-0b3588e7-1013-4a8f-9a5a-a0d16a4a4a2f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI3NzA0NTksIm5iZiI6MTc2Mjc3MDE1OSwicGF0aCI6Ii82NTY5NjA5NS81MTA1ODk5NjMtMGIzNTg4ZTctMTAxMy00YThmLTlhNWEtYTBkMTZhNGE0YTJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEwVDEwMjIzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1NjU0OTA5NzE1MTA4OTgyNDE1MGQxNTU4Y2EwMzdmNGY5ZTE5MDlhOTE5NzhhMjg3M2Y3MWM1YzJlNDNhNGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ymm3dq_YSoA-Ix97y3pIe2bZF_QRbkJm4krSVO2OYYs">
> (3) Click OK to call the backend interface to transfer the project 2. Backend: Interface modification of user related table data to ensure data consistency t_ds_project t_ds_project_parameter t_ds_project_preference t_ds_relation_project_user t_ds_workflow_definition t_ds_workflow_definition_log t_ds_task_definition t_ds_task_definition_log t_ds_schedules t_ds_datasource t_ds_relation_datasource_user t_ds_k8s_namespace t_ds_relation_namespace_user
> 
> Option 2 Increase owner permissions Advantages: No modification of table associated data Disadvantage: Requires compatibility with historical data or permission logic compatible with historical logic
> 
> 1. Front end:
>    Add the "Grant Owner Permissions" button for authorization
> 
> <img alt="Image" width="1598" height="800" src="https://private-user-images.githubusercontent.com/65696095/510590100-45f7abf7-76bf-4e77-8d8c-c6640b38f08e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjI3NzA0NTksIm5iZiI6MTc2Mjc3MDE1OSwicGF0aCI6Ii82NTY5NjA5NS81MTA1OTAxMDAtNDVmN2FiZjctNzZiZi00ZTc3LThkOGMtYzY2NDBiMzhmMDhlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEwVDEwMjIzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiNDY0OGQzN2Y3NzVkZGUzOTMwMGM0MjljNDhhMzExNGZjZTYwNzVhYzU0Y2ZmOTZkYzdjZGNiOTgzZTk4ZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mdPmNINPx7VFhGIqxKxPebOTQOVPwYjwDS1vh5viBRQ">
> 2. Backend:
>    (1) On the basis of the previous permissions, add the owner permission OWNER_PERMISSION=99 to have all permissions. Replace the project creator and modify the verification logic of the previous project creator
>    (2) Transfer authorization logic before reusing permissions, transfer project permissions, increase owner permissions for target users, and revoke all permissions for source users; The source user and target user belong to the same tenant

I prefer the second approach.

**Comment by niumy0701** [comment]
> > > > [@niumy0701](https://github.com/niumy0701) Can we only add a status in User to mark the user is deleted? It's hard to real delete the user, since we need to change a lot of data, e.g. create_user, update_user.
> > > 
> > > 
> > > [@ruanwenjun](https://github.com/ruanwenjun) Equivalent to logically deleting a user, how to transfer the projects associated with the user to others for better handling
> > 
> > 
> > If we don't really delete the user in DB, then we only need to consider how to change the important resource owner, we can ignore some instance create_user/owner/executor.
> 
> I think we don't really delete the user in DB is a better way, but we still need to support transfer project ownership.

Yes, deleting users should be another issue. Before deleting users, it is necessary to transfer the project associated with them; Or simply transferring the project without deleting the user

**Comment by niumy0701** [comment]
> > > [@niumy0701](https://github.com/niumy0701) Can we only add a status in User to mark the user is deleted? It's hard to real delete the user, since we need to change a lot of data, e.g. create_user, update_user.
> > 
> > 
> > [@ruanwenjun](https://github.com/ruanwenjun) Equivalent to logically deleting a user, how to transfer the projects associated with the user to others for better handling
> 
> If we don't really delete the user in DB, then we only need to consider how to change the important resource owner, we can ignore some instance create_user/owner/executor.

Yes, as mentioned in the first solution, it is necessary for users to modify the permission verification association table
t_ds_project
t_ds_relation_project_user
t_ds_datasource
t_ds_relation_datasource_user
t_ds_k8s_namespace
t_ds_relation_namespace_user
If considering table data consistency, other tables also need to be modified, but it is not necessary
@ruanwenjun Do you lean towards Option One？
